image/image_test.go

```diff
@@ -0,0 +1,301 @@
+        "Labels": {},
+        "OnBuild": null,
+        "OpenStdin": false,
+        "StdinOnce": false,
+        "Tty": false,
+        "User": "",
+        "Volumes": null,
+        "WorkingDir": ""
+    },
+    "container": "764ef4448baa9a1ce19e4ae95f8cdd4eda7a1186c512773e56dc634dff208a59",
+    "container_config": {
+        "AttachStderr": false,
+        "AttachStdin": false,
+        "AttachStdout": false,
+        "Cmd": [
+            "/bin/sh",
+            "-c",
+            "#(nop) CMD [\"sh\"]"
+        ],
+        "Domainname": "",
+        "Entrypoint": null,
+        "Env": [
+            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
+        ],
+        "Hostname": "55cd1f8f6e5b",
+        "Image": "sha256:e732471cb81a564575aad46b9510161c5945deaf18e9be3db344333d72f0b4b2",
+        "Labels": {},
+        "OnBuild": null,
+        "OpenStdin": false,
+        "StdinOnce": false,
+        "Tty": false,
+        "User": "",
+        "Volumes": null,
+        "WorkingDir": ""
+    },
+    "created": "2016-06-23T23:23:37.198943461Z",
+    "docker_version": "1.10.3",
+    "history": [
+        {
+            "created": "2016-06-23T23:23:36.73131105Z",
+            "created_by": "/bin/sh -c #(nop) ADD file:9ca60502d646bdd815bb51e612c458e2d447b597b95cf435f9673f0966d41c1a in /"
+        },
+        {
+            "created": "2016-06-23T23:23:37.198943461Z",
+            "created_by": "/bin/sh -c #(nop) CMD [\"sh\"]",
+            "empty_layer": true
+        }
+    ],
+    "os": "linux",
+    "rootfs": {
+        "diff_ids": [
+            "sha256:8ac8bfaff55af948c796026ee867448c5b5b5d9dd3549f4006d9759b25d4a893"
+        ],
+        "type": "layers"
+    }
+}
+`
+)
+
+var (
+	refStr = `{"digest":"<manifest_digest>","mediaType":"application/vnd.oci.image.manifest.v1+json","size":<manifest_size>}`
+
+	manifestStr = `{
+    "annotations": null,
+    "config": {
+        "digest": "<config_digest>",
+        "mediaType": "application/vnd.oci.image.config.v1+json",
+        "size": <config_size>
+    },
+    "layers": [
+        {
+            "digest": "<layer_digest>",
+            "mediaType": "application/vnd.oci.image.layer.tar+gzip",
+            "size": <layer_size>
+        }
+    ],
+    "mediaType": "application/vnd.oci.image.manifest.v1+json",
+    "schemaVersion": 2
+}
+ `
+)
+
+func TestValidateLayout(t *testing.T) {
+	root, err := ioutil.TempDir("", "oci-test")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(root)
+
+	err = os.MkdirAll(filepath.Join(root, "blobs", "sha256"), 0700)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = os.MkdirAll(filepath.Join(root, "refs"), 0700)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	desc, err := createLayerFile(root)
+	if err != nil {
+		t.Fatal(err)
+	}
+	manifestStr = strings.Replace(manifestStr, "<layer_digest>", desc.Digest, -1)
+	manifestStr = strings.Replace(manifestStr, "<layer_size>", strconv.FormatInt(desc.Size, 10), -1)
```

Comment:
No need to use `-1` for these when you know there will only be a single replacement.  And I'm not sure I see the benefit to using templates like this vs. serializing the JSON from the Go types.  In #159 I've floated [`PutJSON`][1] to make adding a CAS blob from a Go type straightforward.  For an example of using it to construct an image, see [here][2].

[1]: https://github.com/opencontainers/image-spec/pull/159/commits/89a36d8af3d17ae72f5f5a8b56fb2562d360bce0#diff-794eda1866432863c39ee59e66210d78R25
[2]: https://github.com/wking/image-spec/commit/2886c655edec1a22d2b246324d5a00c645e57a9f#diff-0e4b5ebdd595bebbfb93671dc3d44ebbR52